### ISSUE TITLE
i18n tag casing

### ISSUE THREAD

**Issue by 747** [ddnexus/pagy#227]
I think pagy currently distinguishes the casing of language tags as argument in the `i18n` module. As I found in the 3.0 release note:
> #### Refactoring of i18n
> - The zh-cn and zh-hk were named incorrectly. They have been properly renamed as zh-CN and zh-HK. If you used them, you should update the Pagy::I18n.load statement.

However, the BCP 47 standard says:
> **2.1.1.  Formatting of Language Tags**
>
> At all times, language tags and their subtags, including private use
> and extensions, are to be treated as case insensitive: there exist
> conventions for the capitalization of some of the subtags, but these
> MUST NOT be taken to carry meaning.
>
> Thus, the tag "mn-Cyrl-MN" is not distinct from "MN-cYRL-mn" or "mN-
> cYrL-Mn" (or any other combination), and each of these variations
> conveys the same meaning: Mongolian written in the Cyrillic script as
> used in Mongolia.

So, I'd rather expect the library to accept case-insensitive language tags for the user input, and take care of normalization internally. I assume that this will not cost too much performance.

**Comment by ddnexus** [comment]
Interesting!

I didn't check the actual code, but I believe a change to accept case-insensitive locale keys will likely have no impact on performance, since it will change loading time, not affecting the run time.

I will take a look at it ASAP. Thank you.

**Comment by ddnexus** [comment]
@747 I looked at the code and reasoned about that, and let's analyze your request to make the user input case insensitive.

It looks like the user input in pagy boils down to just the `@pagy_locale` backend variable. Indeed here are the other aspects of the locale:

1. it is more than acceptable that the `Pagy::I18n` configuration - being a configuration - would require the user to stick to the pagy requirements (see next point about the requirements)

2. the pagy internal convention should be always using lowercase in configuration and files. That will support an easy and efficient way to handle the input (just downcase it) and would require just some minimal changes in dictionary files naming and locale (again just downcased names).

At this point it looks like the only thing to do on the `@pagy_locale` would be downcasing it. But even if it is a minimal performance hit, it does affect run-time, not load-time. So why should we add an internal downcasing for every translation, when the user can just add it to its code if s/he deem it necessary?

I don't think anybody would mind writing `@pagy_locale = params[:locale].downcase` only in case they cannot control the case of the param themselves. For anybody else (the wast majority), hardcoded downcased locale strings would be a better option to avoid downcasing each time a string get translated.

Considering that pagy is a consumer of `I18n`, not an `I18n` gem, I think that the flexibility you are requiring should be possible, but not mandatory. 

In order to allow easy downcasing, I think an internal downcased normalization of files and dictionary keys is in order, along with some documentation notes about the subject.

Thoughts?

**Comment by ddnexus** [comment]
I missed one consideration: legacy apps could be broken by the downcase change, so we should think about a solution to avoid that efficiently.

**Comment by 747** [comment]
@ddnexus Thank you for your detailed analysis. Firstly, I wasn't very clear that my original point was exclusively about `Pagy::I18n`, since I only noticed it via my old script and I haven't use `I18n` gem (if anything, I prefer `R18n` much to it). But thank you for let us be aware of potential interaction problem with `I18n`. Does it mean `Pagy` has a possibility to be fed locale keys directly by `I18n`?

If nothing, I support that you make a convention that all keys must be in lower case in the light of performance, as it's some sort of steady practice. In this case, this fact should be noted somewhere in the documentation.

> I missed one consideration: legacy apps could be broken by the downcase change, so we should think about a solution to avoid that efficiently.

Currently, if I miss nothing, the only ones affected are `zh-CN`, `zh-HK`, and `zh-TW` that exist in the official repository. If so, how about putting them into a temporary substitution table, and deprecate at some point in the future?

**Comment by ddnexus** [comment]
> Does it mean Pagy has a possibility to be fed locale keys directly by I18n?

See the [I18n extra](https://ddnexus.github.io/pagy/extras/i18n)

>  how about putting them into a temporary substitution table, and deprecate at some point in the future?

Not sure about what you mean with "substitution table". If you mean a sort of internal map, yes, that is one possibility, but I have to evaluate how efficient it would be.

The deprecation will be done at the same time of the change.

**Comment by 747** [comment]
I get it. So, if I understand correctly, `I18n` may eat `Pagy`'s names, but not vice versa, right?
https://github.com/ddnexus/pagy/blob/master/lib/pagy/extras/i18n.rb

As far as I've found `I18n` employs IETF style normalization, so we don't have to worry about them.
https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/locale/tag/rfc4646.rb
Should `Pagy` need to process `I18n`'s output, `pt-br` would be a problem either way (because `br` is a country code).

> The deprecation will be done at the same time of the change.

Oops, of course I mean something like "end of support".

**Comment by ddnexus** [comment]
I don't get what you mean with "pagy's names" nor with "we don't have to worry". Please, could you explain the whole previous message? :)

**Comment by 747** [comment]
Since I'm not familiar how to call such thing correctly in the Pagy context, let me call those strings that describe a certain language/locale environment like `en`, `pt-br`, `zh-CN` as _language tags_, as I usually do.

When I looked at this file via your previous link https://github.com/ddnexus/pagy/blob/master/lib/pagy/extras/i18n.rb:
``` ruby
# See the Pagy documentation: https://ddnexus.github.io/pagy/extras/i18n
# encoding: utf-8
# frozen_string_literal: true

class Pagy
  # Use ::I18n gem
  module Frontend

    ::I18n.load_path += Dir[Pagy.root.join('locales', '*.yml')]

    Pagy::I18n.clear.instance_eval { undef :load; undef :t } # unload the pagy default constant for efficiency

    alias_method :pagy_without_i18n, :pagy_t
    def pagy_t_with_i18n(*args) ::I18n.t(*args) end
    alias_method :pagy_t, :pagy_t_with_i18n

  end
end
```
I assumed that this module does only two things in relation to language tag.
1. `::I18n.load_path += Dir[Pagy.root.join('locales', '*.yml')]`
  &mdash; giving `I18n` file names that contain language tags as part of them
2. `def pagy_t_with_i18n(*args) ::I18n.t(*args) end`
  &mdash; giving user input arguments that might contain language tags

Both are flows from `Pagy` to `I18n`, and there doesn't seem a trace that `Pagy` gets a reflux of language tags that `I18n` has processed. It seems that `Pagy` just delegated the rest of process to `I18n`. (Of course I may be wrong when you or `I18n` makes it feeds back language tag info to `Pagy` somewhere, but it'll be quite black magical so I omitted that possibility.)

As for `I18n`, When I wrote the previous comment I thought that all language tags were automatically normalized by https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/locale/tag/rfc4646.rb in `I18n`, but that turned out not true. The default handling is https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/locale/tag/simple.rb which just passes through coder's input as-is and performs exact string/symbol matching, like `Pagy` currently does. But it still means that there's no "compatibility issue" either if we change the convention on the `Pagy`'s side. Who are affected are probably confined to those who already use existing language tags provided by `Pagy`.

That's what I wanted to say in the previous comment.

**Comment by 747** [comment]
Also, if we abandon the idea of case-insensitive matching of language tags, the current naming practice in https://github.com/ddnexus/pagy/tree/master/lib/locales still seems lacking principle (or predictability). I think we have two easy solutions:

1. conform to [canonical forms](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry)
  &mdash; `pt-br` → `pt-BR`
2. all lower case convention
  &mdash; `zh-CN` → `zh-cn`, `zh-HK` → `zh-hk`, `zh-TW` → `zh-tw`

**Comment by ddnexus** [comment]
Thanks.
Opted for `pt-br` -> `pt-BR` normalization. Included in pushed version 3.7.4.
